cirq/optimizers/two_qubit_decompositions.py

```diff
@@ -16,172 +16,41 @@
-    r's phase is the amount of rotation around M's rotation axis.
-
-    This decomposition can be used to decompose controlled single-qubit
-    rotations into controlled-Z operations bordered by single-qubit operations.
-
-    Args:
-      mat:  The qubit operation as a 2x2 unitary matrix.
-
-    Returns:
-        A 2x2 unitary U, the complex relative phase factor r, and the complex
-        global phase factor g. Applying M is equivalent (up to global phase) to
-        applying U, rotating around the Z axis to apply r, then un-applying U.
-        When M is controlled, the control must be rotated around the Z axis to
-        apply g.
-    """
-    vals, vecs = np.linalg.eig(mat)
-    u = np.conj(vecs).T
-    r = vals[1] / vals[0]
-    g = vals[0]
-    return u, r, g
-
-
-def controlled_op_to_operations(
-        control: ops.QubitId,
-        target: ops.QubitId,
-        operation: np.ndarray,
-        tolerance: float = 0.0) -> List[ops.Operation]:
-    """Decomposes a controlled single-qubit operation into Z/XY/CZ gates.
+def two_qubit_matrix_to_operations(q0: ops.QubitId,
+                                   q1: ops.QubitId,
+                                   mat: np.ndarray,
+                                   allow_partial_czs: bool,
+                                   tolerance: float = 1e-8
+                                   ) -> List[ops.Operation]:
+    """Decomposes a two-qubit operation into Z/XY/CZ gates.
 
     Args:
-        control: The control qubit.
-        target: The qubit to apply an operation to, when the control is on.
-        operation: The single-qubit operation being controlled.
+        q0: The first qubit being operated on.
+        q1: The other qubit being operated on.
+        mat: Defines the operation to apply to the pair of qubits.
+        allow_partial_czs: Enables the use of Partial-CZ gates.
         tolerance: A limit on the amount of error introduced by the
             construction.
 
     Returns:
-        A list of Operations that apply the controlled operation.
+        A list of operations implementing the matrix.
     """
-    u, z_phase, global_phase = single_qubit_op_to_framed_phase_form(operation)
-    if abs(z_phase - 1) <= tolerance:
-        return []
-
-    u_gates = single_qubit_matrix_to_gates(u, tolerance)
-    if u_gates and isinstance(u_gates[-1], ops.ZPowGate):
-        # Don't keep border operations that commute with CZ.
-        del u_gates[-1]
-
-    ops_before = [gate(target) for gate in u_gates]
-    ops_after = protocols.inverse(ops_before)
-    effect = ops.CZ(control, target) ** (cmath.phase(z_phase) / math.pi)
-    kickback = ops.Z(control) ** (cmath.phase(global_phase) / math.pi)
-
-    return list(ops.flatten_op_tree((
-        ops_before,
-        effect,
-        kickback if abs(global_phase - 1) > tolerance else [],
-        ops_after)))
+    kak = linalg.kak_decomposition(mat, linalg.Tolerance(atol=tolerance))
+    # TODO: Clean up angles before returning
```

Comment:
Can you clarify what sort of cleanup remains to be done here? Is it related to exponents not always falling into the [0, 2) interval (see other comment)?

Reply:
This is leftover from the old code. I preserved it. I don't know what it refers to anymore; it may have been talking about the single qubit glut. I've deleted it.